src/algebra/divisibility.lean

```diff
@@ -89,4 +128,67 @@ dvd.elim h (assume c, assume H' : a = 0 * c, eq.trans H' (zero_mul c))
 
 @[simp] theorem dvd_zero (a : α) : a ∣ 0 := dvd.intro 0 (by simp)
 
-end comm_monoid_with_zero
+end monoid_with_zero
+
+/-- Given two elements b, c of an integral domain and a nonzero element a, `a*b` divides `a*c` iff
+  b divides c. -/
+theorem mul_dvd_mul_iff_left [cancel_monoid_with_zero α] {a b c : α}
+  (ha : a ≠ 0) : a * b ∣ a * c ↔ b ∣ c :=
+exists_congr $ λ d, by rw [mul_assoc, mul_right_inj' ha]
+
+/-!
+### Units in various monoids
+-/
+
+namespace units
+
+section monoid
+variables [monoid α] (a b : α) (u : units α)
+
+/-- Elements of the unit group of a monoid represented as elements of the monoid
+    divide any element of the monoid. -/
+@[simp] lemma coe_dvd : ↑u ∣ a := ⟨↑u⁻¹ * a, by simp⟩
+
+/-- In a monoid, an element `a` divides an element `b` iff `a` divides all
+    associates of `b`. -/
+@[simp] lemma dvd_coe_mul : a ∣ b * u ↔ a ∣ b :=
+iff.intro
+  (assume ⟨c, eq⟩, ⟨c * ↑u⁻¹, by rw [← mul_assoc, ← eq, units.mul_inv_cancel_right]⟩)
+  (assume ⟨c, eq⟩, eq.symm ▸ dvd_mul_of_dvd_left (dvd_mul_right _ _) _)
+
+/-- An element of a monoid divides a unit iff the element divides one. -/
+@[simp] lemma dvd_coe : a ∣ ↑u ↔ a ∣ 1 :=
+suffices a ∣ 1 * ↑u ↔ a ∣ 1, by simpa,
+dvd_coe_mul _ _ _
+
+end monoid
+
+/-- In a monoid, an element a divides an element b iff all associates of a divide b.-/
+@[simp] lemma coe_mul_dvd [comm_monoid α] (a b : α) (u : units α) : a * u ∣ b ↔ a ∣ b :=
+iff.intro
+  (assume ⟨c, eq⟩, ⟨c * ↑u, eq.symm ▸ by ac_refl⟩)
+  (assume h, suffices a * ↑u ∣ b * 1, by simpa, mul_dvd_mul h (coe_dvd _ _))
+
+end units
+
+namespace is_unit
+
+lemma mul_right_dvd_of_dvd [monoid α]
+  {a b c : α} (hb : is_unit b) (h : a ∣ c) : a * b ∣ c :=
+begin
+  rcases hb with ⟨b, rfl⟩,
+  rcases h with ⟨c', rfl⟩,
+  use (b⁻¹ : units α) * c',
+  rw [mul_assoc, units.mul_inv_cancel_left]
+end
+
+lemma mul_left_dvd_of_dvd [comm_monoid α]
+  {a b c : α} (hb : is_unit b) (h : a ∣ c) : b * a ∣ c :=
+begin
+  rcases hb with ⟨b, rfl⟩,
+  rcases h with ⟨c', rfl⟩,
+  use (b⁻¹ : units α) * c',
+  rw [mul_comm (b : α) a, mul_assoc, units.mul_inv_cancel_left]
+end
+
+end is_unit
```

Comment:
I'd prefer to have a very similar API for `units` and `is_unit`, probably with `mul_left/right` instead of `mul_coe/coe_mul`.

Reply:
I've given this a shot.

Reply:
Oyyy, it appears that this whole section, which I lifted from ring/basic, is already in associated, with slightly different names, outside of the units/is_unit namespaces.